packages/table/src/interactions/selectable.tsx

```diff
@@ -110,142 +124,92 @@ export class DragSelectable extends React.Component<IDragSelectableProps, {}> {
+            this.handleAddDisjointSelection(region);
         } else {
-            onSelection([region]);
-            focusSelectionIndex = 0;
+            this.handleReplaceSelection(region);
         }
 
-        const fullFocusCellCoordinates: IFocusedCellCoordinates = {
-            col: focusCellCoordinates.col,
-            focusSelectionIndex,
-            row: focusCellCoordinates.row,
-        };
-        this.props.onFocus(fullFocusCellCoordinates);
-
         return true;
     }
 
     private handleDragMove = (event: MouseEvent, coords: ICoordinateData) => {
-        const nextSelectedRegions = this.getDragSelectedRegions(event, coords);
-        if (nextSelectedRegions == null) {
+        const {
+            allowMultipleSelection,
+            focusedCell,
+            locateClick,
+            locateDrag,
+            selectedRegions,
+            selectedRegionTransform,
+        } = this.props;
+
+        let region = allowMultipleSelection
+            ? locateDrag(event, coords, /* returnEndOnly? */ this.didExpandSelectionOnActivate)
+            : locateClick(event);
+
+        if (!Regions.isValid(region)) {
             return;
+        } else if (selectedRegionTransform != null) {
+            region = selectedRegionTransform(region, event, coords);
         }
+
+        const nextSelectedRegions = this.didExpandSelectionOnActivate
+            ? this.expandSelectedRegions(selectedRegions, region, focusedCell)
+            : Regions.update(selectedRegions, region);
+
         this.maybeInvokeSelectionCallback(nextSelectedRegions);
 
-        if (!this.props.allowMultipleSelection) {
-            // have the focused cell follow the selected region
-            const mostRecentRegion = nextSelectedRegions[nextSelectedRegions.length - 1];
-            const focusCellCoordinates = Regions.getFocusCellCoordinatesFromRegion(mostRecentRegion);
-            const fullFocusCellCoordinates: IFocusedCellCoordinates = {
-                col: focusCellCoordinates.col,
-                focusSelectionIndex: nextSelectedRegions.length - 1,
-                row: focusCellCoordinates.row,
-            };
-            this.props.onFocus(fullFocusCellCoordinates);
+        if (!allowMultipleSelection) {
+            // move the focused cell with the selected region
+            const lastIndex = nextSelectedRegions.length - 1;
+            const mostRecentRegion = nextSelectedRegions[lastIndex];
+            this.invokeOnFocusCallbackForRegion(mostRecentRegion, lastIndex);
         }
     }
 
-    private handleDragEnd = (event: MouseEvent, coords: ICoordinateData) => {
-        const nextSelectedRegions = this.getDragSelectedRegions(event, coords);
-        if (nextSelectedRegions == null) {
-            return;
-        }
-        this.maybeInvokeSelectionCallback(nextSelectedRegions);
-        BlueprintUtils.safeInvoke(this.props.onSelectionEnd, nextSelectedRegions);
+    private handleDragEnd = () => {
```

Comment:
We were doing duplicate work on `dragEnd` and on `click` that was already done on `mousedown` and/or on `mousemove`.